### PR TITLE
refactor(optimizer): finalize optimizer schema and backend handling

### PR DESCRIPTION
**Pull request by OutisLi**
## Summary by CodeRabbit

* **New Features**
  * Added a top-level, configurable optimizer section; PyTorch: Adam/AdamW/LKF/AdaMuon/HybridMuon, TensorFlow/Paddle: Adam.
  * Exposes Adam-style hyperparameters (beta1, beta2, weight_decay) for relevant optimizers.

* **Documentation**
  * Training docs updated with optimizer examples and framework-specific guidance.

* **Backward Compatibility**
  * Legacy optimizer fields are auto-migrated to the new format; tests and examples updated to use the new config.

### CODE REVIEW

File: deepmd/pt/train/training.py [hunk 1]
```diff
@@ -157,7 +158,24 @@ def __init__(
...
+                default_weight_decay = 0.0
+            elif opt_type == "LKF":
+                default_adam_beta2 = 0.95
+                default_weight_decay = 0.001
+            elif opt_type == "AdaMuon":
+                default_adam_beta2 = 0.95
+                default_weight_decay = 0.001
+            elif opt_type == "HybridMuon":
+                default_adam_beta2 = 0.95
+                default_weight_decay = 0.001
```
Comment (iProzd) [inline deepmd/pt/train/training.py]: These default values should only appear in argcheck?

File: deepmd/utils/argcheck.py [hunk 2]
```diff
@@ -3396,183 +3680,8 @@ def training_args(
             doc=doc_only_pd_supported + doc_acc_freq,
         ),
     ]
-    variants = [
-        Variant(
-            "opt_type",
```
Comment (iProzd) [inline deepmd/utils/argcheck.py:3623]: This pr seems a breaking change. Is it possible to make old config compatible?

**Review by iProzd — CHANGES_REQUESTED** [review]

File: deepmd/utils/argcheck.py (hunk 2 above)
Comment (njzjz) [reply deepmd/utils/argcheck.py:3623]: Have we discussed this change? Btw, there should be no breaking change.

File: deepmd/pt/train/training.py [hunk 3]
```diff
@@ -157,24 +158,33 @@ def __init__(
...
+            Extract optimizer parameters.
+
+            Note: Default values are already filled by argcheck.normalize()
+            before this function is called, so we use params.get() without
+            fallback defaults here.
+            """
+            opt_type = params.get("type", "Adam")
+            if opt_type not in ("Adam", "AdamW", "LKF", "AdaMuon", "HybridMuon"):
+                raise ValueError(f"Not supported optimizer type '{opt_type}'")
             opt_param = {
```
Comment (wanghan-iapcm) [inline deepmd/pt/train/training.py]: I would suggest not formatting `opt_param` here. For example, when using the Adam optimizer, the KF parameters are not applicable and therefore should not be retrieved from the `params` directory.

one may simply assign params to opt_param

File: deepmd/pt/train/training.py [hunk 4]
```diff
@@ -712,20 +722,38 @@ def warm_up_linear(step: int, warmup_steps: int) -> float:
...
-                    fused=False if DEVICE.type == "cpu" else True,
-                )
-            else:
-                self.optimizer = torch.optim.AdamW(
-                    self.wrapper.parameters(),
-                    lr=self.lr_exp.start_lr,
-                    weight_decay=float(self.opt_param["weight_decay"]),
-                    fused=False if DEVICE.type == "cpu" else True,
-                )
+        if self.opt_type == "Adam":
```
Comment (wanghan-iapcm) [inline deepmd/pt/train/training.py]: here get parameters from the `self.opt_param` based on the type of the optimizer.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

File: deepmd/pd/train/training.py [hunk 5]
```diff
@@ -120,6 +120,7 @@ def __init__(
         self.restart_training = restart_model is not None
         model_params = config["model"]
         training_params = config["training"]
+        optimizer_params = config.get("optimizer", {})
```
[the remark this answers is not in the document]
Comment (OutisLi) [reply deepmd/pd/train/training.py:123]: the normalizer in argcheck will prevent this issue

File: deepmd/pd/train/training.py [hunk 6]
```diff
@@ -157,14 +158,17 @@ def __init__(
...
-            }
+            """
+            Extract optimizer parameters.
+
+            Note: Default values are already filled by argcheck.normalize()
+            before this function is called.
+            """
+            opt_type = params.get("type", "Adam")
+            if opt_type != "Adam":
+                raise ValueError(f"Not supported optimizer type '{opt_type}'")
```
[the remark this answers is not in the document]
Comment (OutisLi) [reply deepmd/pd/train/training.py:169]: just like before. These two are quite alike, whether to raise during training init or config loading

File: deepmd/utils/argcheck.py [hunk 7]
```diff
@@ -2742,6 +2742,302 @@ def _check_lr_args(data: dict[str, Any]) -> bool:
...
+            float,
+            optional=True,
+            default=0.999,
+            doc=doc_only_pt_supported + doc_adam_beta2,
+        ),
+        Argument(
+            "weight_decay",
+            float,
+            optional=True,
+            default=0.0,
```
Comment (iProzd) [inline deepmd/utils/argcheck.py]: Default value of `weight_decay` for adamW in DPA3 was 0.001

**Review by iProzd — CHANGES_REQUESTED** [review]
Add default adamW params in dpa3 examples?
Others LGTM.

**Review by njzjz — APPROVED** [review]

File: deepmd/utils/compat.py [hunk 8]
```diff
@@ -457,4 +457,92 @@ def is_deepmd_v1_input(jdata: dict[str, Any]) -> bool:
...
+    return jdata
+
+
+def convert_optimizer_to_new_format(
+    jdata: dict[str, Any], warning: bool = True
+) -> dict[str, Any]:
+    """Convert old optimizer format (in training section) to new format (separate optimizer section).
+
+    Old format: optimizer parameters (opt_type, kf_blocksize, etc.) in training section.
+    New format: separate optimizer section with type field.
```
Comment (wanghan-iapcm) [inline deepmd/utils/compat.py]: - "old" and "new" may not clear enough to indicate to which version the optimizer section is compatible. 
- Currently the "old" indicates <= v3.1, while if we decide to accept this pr, we need to bump version to 3.2.
- the name of this method may be designed as convert_optimizer_v31_to_v32

File: deepmd/pt/train/training.py [hunk 9]
```diff
@@ -328,7 +320,7 @@ def get_lr(lr_params: dict[str, Any]) -> BaseLR:
                     self.optim_dict[model_key]
                 )
```
Comment (wanghan-iapcm) [inline deepmd/pt/train/training.py]: it seems that for multi-task training, the "optim_dict" section is still assume to be in the training section, am I correct? 
We may want to make the convention compatible with single task training.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

File: deepmd/utils/compat.py [hunk 10]
```diff
@@ -457,4 +457,92 @@ def is_deepmd_v1_input(jdata: dict[str, Any]) -> bool:
         jdata = deprecate_numb_test(jdata, warning, dump)
 
     jdata = migrate_training_warmup(jdata, warning=warning)
+    jdata = convert_optimizer_v31_to_v32(jdata, warning=warning)
+    return jdata
+
+
+def convert_optimizer_v31_to_v32(
```
Comment (wanghan-iapcm) [inline deepmd/utils/compat.py:464]: please add a UT for the converter. Thanks!

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

**Review by iProzd — APPROVED** [review]

**Review by wanghan-iapcm — APPROVED** [review]
